vignettes/oauth.Rmd

```diff
@@ -0,0 +1,469 @@
+browseURL(tpapp2$url("/login"))
+```
+
+```{r include = FALSE}
+login2 <- oauth2_login(tpapp2$url("/login"))
+```
+
+```{r echo = FALSE}
+cat(rawToChar(login2$token_response$content))
+```
+
+The fake third party app also has an endpoint to return the saved tokens:
+
+```{r}
+httr::content(httr::GET(tpapp2$url("/locals")))
+```
+
+Now the third-party app can get data on your behalf (the whole goal of OAuth!) --- it could also post data on your behalf if the resource app had endpoints for that.
+
+For example the `/data` endpoint of the third party app queries the resource app and needs authentication.
+If you run the following without the OAuth dance, your access is denied.
+But now it works fine:
+
+```{r}
+resp_data <- httr::GET(tpapp2$url("/data"))
+resp_data
+httr::content(resp_data, as = "text")
+```
+
+In real life, access by the third-party app might be limited to some scopes, but the fake apps shipped with webfakes do not handle that.
+
+## The fake resource server and httr
+
+When you use httr's OAuth tool, there's some gymnastics happening as R is playing the role of a third-party app via httr and httpuv (to listen to the redirect URI).
+
+### OAuth2.0 app creation & registration
+
+What's crucial here is setting `httr::oauth_callback()` as redirect URI, which is what you do when creating an app for an R package that uses OAuth2.0 to authenticate to a resource server.
+
+```{r httr}
+url3 <- paste0(
+  regi_url,
+  "?name=3P%20app2",
+  "&redirect_uri=", httr::oauth_callback()
+)
+reg_resp3 <- httr::GET(url3)
+reg_resp3
+regdata3 <- httr::content(reg_resp3)
+regdata3
+```
+
+Now we set the registration data on the third-party app.
+
+```{r httr-dance}
+app <- httr::oauth_app(
+  "3P app2",
+  key = regdata3$client_id[[1]],
+  secret = regdata3$client_secret[[1]],
+  redirect_uri = httr::oauth_callback()
+)
+
+endpoint <- httr::oauth_endpoint(
+  authorize = auth_url,
+  access = toke_url
+)
+```
+
+Now we can launch the token creation.
+
+```{r httr-dance2}
+token <- oauth2_httr_login(
+  httr::oauth2.0_token(endpoint, app, cache = FALSE)
+)
+```
+
+```{r refresh}
+token
+```
+
+Without the token, the query to the resource server fails:
+
+```{r}
+httr::GET(rsapp$url("/data"))
+```
+
+With the token, it is successful.
+httr also automatically refreshes the token if needed.
+
+```{r}
+httr::content(
+  httr::GET(rsapp$url("/data"), config = token),
+  as = "text"
+)
+```
+
+## Advanced topics
+
+### Applications
+
+With these apps, or only the resource server app, you can now test your code that helps users create and store an OAuth2.0 token.
+
+Like all webfakes apps, OAuth2.0 apps are extensible: you can add your own endpoints and middleware to it.
+E.g. here we add logging via `mw_log()` and a new endpoint.
+
+```{r}
+rsapp2 <- oauth2_resource_app(
+  refresh_duration = .Machine$integer.max,
+  access_duration = 10L
+)
+logfile <- tempfile("oauth-rs-", fileext = ".log")
+rsapp2$use(mw_log(stream = logfile), .first = TRUE)
```

Comment:
can you add a comment about why it's important to use `.first` here?

Reply:
I don't really want to go into the details, because I want to change this later.